### PR TITLE
Add .configuration to Grape::API (fixes #1906)

### LINKED ISSUE #1906: Dependency injection?
Hi! Sorry if this is a stupid question, but I've searched the documentation and it's very unclear to me how I can pass objects from outside Grape::API's scope through, say, a constructor, for dependency injection purposes?

I've tried using a bog-standard constructor, but a class derived from Grape::API doesn't seem to like being initialized:

```ruby
class Foo < Grape::API
  def initialize(baz)
    @baz = baz
  end

  get :bar do
    @baz
  end
end

run Foo.new('hello') # ArgumentError
```

If I remove the `.new` call so it's just `run Foo` the constructor isn't even called, so `@baz` is obviously nil.

It seems `configuration` would be able to do this but it's _incredibly_ unclear to me how to use it outside of `mount`, or, for that matter, when initialising the object which is what I want to do :)

There's also another, less related question. Are routes supposed to use symbols or strings? Or either? If it's either, is there a convention? That also isn't clear to me.

Thank you and I hope I'm not being too dense lol :p

### PR DESCRIPTION
**Pull request by unleashy**
Still work in progress... includes an intentionally failing test for the inheritance of config to mounted APIs, which I am totally unsure of how to implement and would really like some help with :)

Also if the method name is appropriate. I prefer `with_configuration(hash)` but @myxoh thought of `with(configuration: hash)`.

Todo list:
- [x] ~~inheritance of config on mounted endpoints~~
- [x] documentation
- [x] changelog

Thanks!

### CODE REVIEW

File: spec/grape/api_spec.rb [hunk 1]
```diff
@@ -3754,6 +3754,42 @@ def before
     end
   end
 
+  describe '.with_configuration' do
+    it 'passes configuration from outside' do
+      subject.with_configuration(hello: 'hello', bread: 'bread')
```
[the remark this answers is not in the document]
Comment (dblock) [reply spec/grape/api_spec.rb]: I don't think you can, but you should be able to. What I am saying is that is a more standard way of doing configuration in Ruby. Check out https://github.com/slack-ruby/slack-ruby-client/blob/master/lib/slack/config.rb for example.

[the remark this answers is not in the document]
Comment (unleashy) [reply spec/grape/api_spec.rb]: Ah, that clears it up, thank you! I'll try my hand at it.

**Comment by dblock** [comment]
I like this, exposes `configure` with no other api changes. Document it?

**Comment by dblock** [comment]
@unleashy Let's finish this? See comments above, bring this to a close?

**Comment by unleashy** [comment]
Sorry for the delay, I have been busy. I'm kinda sad that I couldn't implement config inheritance but it's ok.

File: lib/grape/api.rb [hunk 2]
```diff
@@ -42,6 +42,12 @@ def override_all_methods!
         end
       end
 
+      def with_configuration(config)
```
[the remark this answers is not in the document]
Comment (myxoh) [reply lib/grape/api.rb]: Sorry for the delay here. If you want I can help on implementing `merge` and `merge!` on `LazyHash`

**Comment by myxoh** [comment]
Hi, sorry for the delay on commenting on this PR. Config inheritance is an interesting one but it might have unintended behaviour, as for the time being the `configuration` was meant to be specific for the API endpoints.
One way I did do inheritance one I thought that was appropiate was
```ruby
class Parent < Grape::API
  mount Child, with: configuration
end

class Child < Grape::API
  get configuration[:key]
    'response
  end
end

class Base <Grape::API
  mount Parent, with: {key: 'value'}
end
```

**Comment by unleashy** [comment]
@myxoh yeah I realised I could do that so it's not a big deal actually hehe.

**Comment by dblock** [comment]
It looks like what we document in README is misleading then because it seems to imply that you are configuring a specific API? Let's explain how it inherits?

@myxoh Anything else you think this PR needs to merge?

**Comment by myxoh** [comment]
Should we just merge this @dblock?

**Comment by dblock** [comment]
I hit merge.

**Comment by dblock** [comment]
Thanks for your solid work @unleashy!

**Comment by unleashy** [comment]
Thank you all :heart_decoration:

**Pull request merged**